### PR TITLE
Nano: Change the API of keras optmizer and quantize

### PR DESCRIPTION
**Pull request by MeouSker77**
## Description

Change the API of keras `InferenceOptimizer` and `quantize`, now is similar to `keras.Model.fit`

### 1. Why the change?

`keras.Model.fit` use `x` and `y` as input dataset, where `x` is tensors or ndarrays of feature, `y` is tensors or ndarrays of label, it also accepts `x` as a `tf.data.Dataset` returning both feature and label. This PR changes current `calib_dataset` parameter to `x` and `y` to be consistent with `keras.Model.fit`

### 2. User API changes

Before:
```python
# quantize
from bigdl.nano.tf.keras import Model

model = Model()
model.quantize(accelerator=..., calib_dataset=train_dataset)

...

# optimize
from tensorflow.keras import Model
from bigdl.nano.tf.keras import InferenceOptimizer

model = Model()
opt = InferenceOptimizer()
opt.optimize(model=model, calib_dataset=train_dataset)
```

Now:
```python
# quantize
from bigdl.nano.tf.keras import Model

model = Model()
model.quantize(accelerator=..., x=train_dataset)
# or
model.quantize(accelerator=..., x=train_features, y=train_labels)

...

# optimize
from tensorflow.keras import Model
from bigdl.nano.tf.keras import InferenceOptimizer

model = Model()
opt = InferenceOptimizer()
opt.optimize(model=model, x=train_dataset)
# or
opt.optimize(model=model, x=train_features, y=train_labels)
```

Before `calib_dataset` must be a dataset which returns tuples of (feature, label)

Now `x` and `y` can be a numpy array or tensor, or `x` is a dataset which returns tuples of (feature, label), i. e.

        :param x: Input data which is used for training. It could be:
                  | 1. a Numpy array (or array-like), or a list of arrays (in case the model
                  | has multiple inputs).
                  |
                  | 2. a TensorFlow tensor, or a list of tensors (in case the model has
                  | multiple inputs).
                  |
                  | 3. an unbatched tf.data.dataset. Should return a tuple of (inputs, targets).

                  X will be used as calibration dataset for Post-Training Static Quantization (PTQ),
                  as well as be used for generating input_sample to calculate latency.
                  To avoid data leak during calibration, please use training dataset.
        :param y: Target data. Like the input data x, it could be either Numpy array(s) or
                  TensorFlow tensor(s). It should be consistent with x (you cannot have Numpy inputs
                  and tensor targets, or inversely). If x is a dataset, y will be ignored (since
                  targets will be obtained from x).

Besides, rename the parameter `input_sample` of `model.trace(...)` to `input_spec`

### 3. Summary of the change 

1. Change the API of keras optmizer and quantize: `calib_dataset` to `x` and `y`
2. Rename the parameter `input_sample` of `model.trace(...)` to `input_spec`
3. Infer input shape when tracing onnx model in `model.trace` instead of in `model.quantize` or `opt.optimize`
4. Better `tensors_to_numpy` function
5. Change and add all related UT

### 4. How to test?
- [ ] N/A
- [x] Unit test
- [ ] Application test
- [ ] Document test
- [ ] ...

### CODE REVIEW

**Comment by jason-dai** [comment]
Do we still use `InferenceOptimizer.optimize/quantize/trace` for TF?

**Comment by MeouSker77** [comment]
In fact, now we are using `InferenceOptimizer.optimize` and `model.quantize/trace`

**Comment by jason-dai** [comment]
> In fact, now we are using `InferenceOptimizer.optimize` and `model.quantize/trace`

Why the change then? In particular, I don't think it's a good idea to provide two type of interfaces for the same type of concept.

**Comment by MeouSker77** [comment]
> Why the change then? In particular, I don't think it's a good idea to provide two type of interfaces for the same type of concept.

We have been using `model.quantize/trace` before adding keras optimizer, we will replace it with `InferenceOptimizer.quantize/trace` in the future

File: python/nano/src/bigdl/nano/tf/keras/inference_utils.py [hunk 1]
```diff
@@ -152,10 +173,7 @@ def quantize(self,
             if isinstance(self, KerasONNXRuntimeModel):     # type: ignore
                 onnx_model = self
             else:
-                spec = tf.TensorSpec((self.input_shape), self.dtype)    # type: ignore
-                onnx_model = self.trace(accelerator='onnxruntime',
-                                        input_sample=spec,
-                                        thread_num=thread_num)
+                onnx_model = self.trace(accelerator='onnxruntime', thread_num=thread_num)
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/tf/keras/inference_utils.py:176]: why don't we need `input_spec` here?

Comment (MeouSker77) [reply python/nano/src/bigdl/nano/tf/keras/inference_utils.py:176]: now it will automatically infer the `input_spec` in`trace`

**Review by TheaperDeng — APPROVED** [review]
LGTM
